### PR TITLE
Introduce Aggregate Scoping, Witnesses, Multi-leveled Injected Variables

### PR DESCRIPTION
**Pull request by rdowavic**
There are some problems in #1667 that this PR will fix. There used to not really be a concept of aggregation scope, but I implement this now.

The TLDR is that you can now:
* Not worry about weird things going on with multiple aggregates in the same rule (like sister aggregates) because we have proper scoping now
*  Inject a variable and it can skip levels, i.e. "reach" into even more inner scopes, not just the next level
* export witnesses into the base scope ONLY, and from ANY level of aggregate nesting.

**Here is some basic outline of what I did with scoping:**
-  The appearance of variables outside of any aggregate are considered part of the "base" scope.
- The appearance of a variable in the target expression of an aggregate renders this variable local to the scope of the aggregate subclause. This just meant materially that if a target variable shared the name of a variable occurring literally anywhere else besides the aggregate subclause that that target expression is connected to (like in the base scope, or even in another aggregate), then I will just give it a unique name. So i.e. you can use the same target variable name in two different aggregates and be assured that there is no relation between them at all like no implication that their values coincide or anything. Also, I need to mention as well that if you have two local variables that have the same name but occur in unrelated aggregates (i.e. there is no ancestral relation from one to the other), then these also get renamed because it won't make sense for them to have any relation to each other. I.e.

n = min x : { B(x, y), C(y, z) },
y = min x : { B(x, y), C(y, z) },

The x, y, z will all get renamed in the second one. And I suppose you have to be careful, here, the y appears GROUNDED in the outer scope (by y = blah) and then it gets used in the first aggregate, so this will be treated as injected. If you want to avoid this you must use different names cuz otherwise how would I distinguish what you want to be local or injected? This is my only mechanism right now.

Oh, by the way, by "local" I mean just not injected and not a witness either. I define these in a second but local is probably just exactly what your intuition already tells you, like a local variable in a function, it cannot interact with the outside.

**Now this is what I've done with the injected variables:**
- If a user writes a rule and it contains a variable called A *grounded* in the base scope, and a variable sharing the same name (A) occurs in the scope of an aggregate subclause, then I consider this "injected", and the name is left as is. The RA query already deals with this by itself and it had always been working as intended, except for when the injected variable appeared more than one level deep (i.e. it appeared in the aggregate subclause of another aggregate subclause). This was just because my/our mechanism to outline the aggregate body was still kind of primitive (MaterializeAggregationQueries) and I hadn't really done a full treatment of nested aggregation yet. So the affect of this PR now is that you are free to write nested aggregation queries that are like this:

.. :- B(x), y = min z : { C(z), z > max w : { D(x, w) } }

You can see that x appears in the atom B(x), and then it also appears in the atom D(x, w), so x is injected all the way in. The way I treat this is like, all levels below this aggregate really contain this injected variable, so they are all parametrised by it, and the variable ends up getting carried up each and every level. It makes all of the aggregates "group-by". Only in even more inner aggregates can the aggregates be considered not group-by aggregates anymore.

-  To clarify, you can inject a variable from any level to any other HIGHER level, so not just the base scope. Like you could have a variable that is local to one aggregate, and then is used again in an inner aggregate, then this is "injected" in that sense and behaves as you would expect. I like to think that the injected variable "anchors" some value, and then calculates another aggregate based on that value, and will calculate the aggregate afresh for every different assignment of that value. It is exactly like "group-by".

**Here is what I did about witnesses:**
- You also now have a witness functionality for minimum and maximum aggregation! What does this mean in practice?

Simple case:

You have an aggregate over like a single relation or something and you just want to figure out what OTHER values are associated with the value that was selected as the min/max. Here you can export the "successful" grounding of the name variable to the head of the rule. This is because the head arguments are considered ungrounded. You can only export a witness to somewhere that is ungrounded. I have not really found a practical use for anywhere to export the witness besides the head. Like maybe in something like a literal like x < 5? I don't know..

Tallest(name) :- _ = min height : { Person(name, height ) }

You can also like, figure out whatever you really want, like what are all the subjects taken by the person with the minimum height? Just do a join in the aggregate subclause, this is how you do it.

SubjectsTakenByTheShortestPerson(subject) :- _ = min height : { Person(name, height), Takes(name, subject)  }.

This witness feature is literally just syntactic sugar. You could already have done this by just replicating all the aggregate subclause literals on the outside and then just filling in the target variable column with the actual aggregate value, i.e. the LHS of the equality literal if that's how you've used it. Or you could just literally copy the entire aggregate argument again. But I suppose this is a lot more succinct.

- There is a caveat to a witness: Although you can export them to the base scope, you CANNOT at the moment export them to an outer scope that is an aggregate subclause. I just felt like I wasn't sure what use case there was for it. I don't know, maybe there is a situation I haven't thought of where that could be useful. So with this restriction in mind, you can still export a witness from ANY level of nesting into the base scope. Like you can write an aggregate, then the aggregate contains an aggregate, and that inner aggregate has a witness in it, and that witness name appears in the head, or ungrounded in the base scope, that is how I "detect" that you are trying to indicate you want to export a witness. So these situations can be dealt with. For example:

E(n, z) :- n = count : { B(x), x < min y : { D(y, z)} }

You can see z is whatever value is associated with the minimum value of the left most column in the D relation. That can go into the head and it is totally fine. I like to call this a reaching witness.

- You can "inject" a witness into an inner scope, so the witness variable is really just acting like a regular local variable in that way. Basically because once I transform out the witnesses, witnesses just look "local" again, and you can do with them all the same things you do with local variables then.

### CODE REVIEW

File: tests/evaluation/magic_aggregates/magic_aggregates.dl [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 // A copy of the `aggregates' test, this time using
 // the magic set transformation
-.pragma "magic-transform" "*"
+//.pragma "magic-transform" "*"
```
Comment (mmcgr) [inline tests/evaluation/magic_aggregates/magic_aggregates.dl:9]: Is this intentional?

File: tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+A(-50, 3).
+A(54, 5).
+
+MinSum(x, y, s) :- s = min w : { A(x, y), w = x + y }.
+MinSumNotReduced(x, y, s) :- s = min x + y : { A(x, y) }.
+
+.output A, MinSum, MinSumNotReduced()
+
+/**
+* Later on I realised I had been ignorant and there were many more interesting witness cases.
```
Comment (mmcgr) [inline tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl:35]: The history of the tests aren't essential to include in the comments. The git log can inform the inquiring historians.

File: tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+F("Reaching witness", result, witness) :- result = count : { B(x, _), x > min y : { D(y, witness) } }.
+
+// test that local variables are scoped (renamed uniquely)
+F("two side-by-side aggregates", res1, res2) :- res1 = count : { D(_, y), y > 1 }, res2 = count : { D(_, y), y < 4 }.
+
+// check that scoping is all gucci!
+G("Serious scoping test", x, y, z) :- x = max y : { D(_, y), y < min y : { B(_, y) } }, y = min y : { H(_, y, _) }, z = min y : { B(y, _), y != 2 }.
+
+/**
+*   Can I have a witness that is then injected into an inner aggregate??
```
Comment (mmcgr) [inline tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl:78]: This comment could be a little more concise

File: tests/evaluation/magic_aggregates/magic_aggregates.dl (hunk 1 above)
Comment (rdowavic) [reply tests/evaluation/magic_aggregates/magic_aggregates.dl:9]: oops sorry about this

**Comment by mmcgr** [comment]
Suggested changes from the linter:
```diff
diff --git a/src/ast/transform/GroundWitnesses.cpp b/src/ast/transform/GroundWitnesses.cpp
index a1bf857..60af2a8 100644
--- a/src/ast/transform/GroundWitnesses.cpp
+++ b/src/ast/transform/GroundWitnesses.cpp
@@ -21,6 +21,7 @@
 #include "souffle/utility/StringUtil.h"
 
 #include <set>
+#include <utility>
 
 namespace souffle::ast::transform {
 
@@ -33,7 +34,7 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
         std::set<std::string> witnesses;
 
         AggregateWithWitnesses(Aggregator* agg, Clause* clause, std::set<std::string> witnesses)
-                : aggregate(agg), originatingClause(clause), witnesses(witnesses) {}
+                : aggregate(agg), originatingClause(clause), witnesses(std::move(witnesses)) {}
     };
 
     std::vector<AggregateWithWitnesses> aggregatesToFix;
@@ -92,9 +93,9 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
             Aggregator* aggregate;
             std::string targetVariable;
             TargetVariableReplacer(Aggregator* agg, std::string target)
-                    : aggregate(agg), targetVariable(target) {}
+                    : aggregate(agg), targetVariable(std::move(target)) {}
             std::unique_ptr<Node> operator()(std::unique_ptr<Node> node) const override {
-                if (Variable* variable = dynamic_cast<Variable*>(node.get())) {
+                if (auto* variable = dynamic_cast<Variable*>(node.get())) {
                     if (variable->getName() == targetVariable) {
                         auto replacement = souffle::clone(aggregate);
                         return replacement;
@@ -104,7 +105,7 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
                 return node;
             }
         };
-        const Variable* targetVariable = dynamic_cast<const Variable*>(agg->getTargetExpression());
+        const auto* targetVariable = dynamic_cast<const Variable*>(agg->getTargetExpression());
         std::string targetVariableName = targetVariable->getName();
         TargetVariableReplacer replacer(agg, targetVariableName);
         for (std::unique_ptr<Literal>& literal : aggregateLiterals) {
```

File: tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl (hunk 2 above)
Comment (rdowavic) [reply tests/evaluation/aggregate_witnesses/aggregate_witnesses.dl:35]: okay

**Comment by mmcgr** [comment]
You mention fixing issue #1617 - Standardise nullptr comparisons. Did you mean #1667 - souffle hangs on aggregator variable re-use?

File: src/ast/transform/SimplifyAggregateTargetExpression.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved
```
Comment (mmcgr) [inline src/ast/transform/SimplifyAggregateTargetExpression.h:3]: Fix date and author as before

File: src/ast/transform/SimplifyAggregateTargetExpression.cpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved
```
Comment (mmcgr) [inline src/ast/transform/SimplifyAggregateTargetExpression.cpp:3]: Fix date and author as before

File: src/ast/transform/SimplifyAggregateTargetExpression.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "ast/transform/SimplifyAggregateTargetExpression.h"
+#include "ast/Argument.h"
+#include "ast/TranslationUnit.h"
+#include "ast/analysis/Aggregate.h"
+#include "ast/transform/GroundWitnesses.h"
+#include "ast/utility/NodeMapper.h"
+#include "ast/utility/Visitor.h"
+
+#include "souffle/utility/StringUtil.h"
```
Comment (mmcgr) [inline src/ast/transform/SimplifyAggregateTargetExpression.cpp:23]:
```suggestion
#include "souffle/utility/StringUtil.h"
#include <memory>
#include <set>
#include <string>
#include <utility>
#include <vector>
```

File: src/ast/transform/MaterializeAggregationQueries.cpp [hunk 7]
```diff
@@ -45,197 +46,305 @@
...
+        headArguments.insert(injectedVarName);
+    }
+    return headArguments;
+}
 
-                for (const auto& lit : clause.getBodyLiterals()) {
-                    const auto* atom = dynamic_cast<const Atom*>(lit);
-                    if (atom == nullptr) {
-                        continue;  // it's not an atom so it can't help ground anything
+// TODO: Deal with recursive parameters with an assert statement.
```
Comment (mmcgr) [inline src/ast/transform/MaterializeAggregationQueries.cpp:109]: Add ownership or issue number to the TODO

File: src/ast/transform/MaterializeAggregationQueries.cpp [hunk 8]
```diff
@@ -45,197 +46,305 @@
...
-                    for (auto const& pair : aggVarMap[&agg]) {
-                        std::string varName = pair.first;
-                        int numOccurrences = pair.second;
-                        aggVarMap[parent[&agg]][varName] -= numOccurrences;
-                    }
+            // 2. Variable must be grounded by this literal.
+            auto singleLiteralClause = mk<Clause>();
+            singleLiteralClause->addToBody(souffle::clone(lit));
+            bool variableGroundedByLiteral = false;
+            for (const auto& ap : analysis::getGroundedTerms(translationUnit, *singleLiteralClause)) {
```
Comment (mmcgr) [inline src/ast/transform/MaterializeAggregationQueries.cpp:195]: `ap`? Can you find a more meaningful name?

**Comment by b-scholz** [comment]
The linter complains about the following idioms:
```
diff --git a/src/ast/transform/GroundWitnesses.cpp b/src/ast/transform/GroundWitnesses.cpp
index a1bf857..60af2a8 100644
--- a/src/ast/transform/GroundWitnesses.cpp
+++ b/src/ast/transform/GroundWitnesses.cpp
@@ -21,6 +21,7 @@
 #include "souffle/utility/StringUtil.h"
 
 #include <set>
+#include <utility>
 
 namespace souffle::ast::transform {
 
@@ -33,7 +34,7 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
         std::set<std::string> witnesses;
 
         AggregateWithWitnesses(Aggregator* agg, Clause* clause, std::set<std::string> witnesses)
-                : aggregate(agg), originatingClause(clause), witnesses(witnesses) {}
+                : aggregate(agg), originatingClause(clause), witnesses(std::move(witnesses)) {}
     };
 
     std::vector<AggregateWithWitnesses> aggregatesToFix;
@@ -92,9 +93,9 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
             Aggregator* aggregate;
             std::string targetVariable;
             TargetVariableReplacer(Aggregator* agg, std::string target)
-                    : aggregate(agg), targetVariable(target) {}
+                    : aggregate(agg), targetVariable(std::move(target)) {}
             std::unique_ptr<Node> operator()(std::unique_ptr<Node> node) const override {
-                if (Variable* variable = dynamic_cast<Variable*>(node.get())) {
+                if (auto* variable = dynamic_cast<Variable*>(node.get())) {
                     if (variable->getName() == targetVariable) {
                         auto replacement = souffle::clone(aggregate);
                         return replacement;
@@ -104,7 +105,7 @@ bool GroundWitnessesTransformer::transform(TranslationUnit& translationUnit) {
                 return node;
             }
         };
-        const Variable* targetVariable = dynamic_cast<const Variable*>(agg->getTargetExpression());
+        const auto* targetVariable = dynamic_cast<const Variable*>(agg->getTargetExpression());
         std::string targetVariableName = targetVariable->getName();
         TargetVariableReplacer replacer(agg, targetVariableName);
         for (std::unique_ptr<Literal>& literal : aggregateLiterals) {
```

**Comment by b-scholz** [comment]
The style checker complains about the following formatting:
```
diff --git a/src/ast/transform/MaterializeAggregationQueries.cpp b/src/ast/transform/MaterializeAggregationQueries.cpp
211index 8315e69..adf0ba7 100644
212--- a/src/ast/transform/MaterializeAggregationQueries.cpp
213+++ b/src/ast/transform/MaterializeAggregationQueries.cpp
214@@ -169,7 +169,7 @@ void MaterializeAggregationQueriesTransformer::groundInjectedParameters(
215                 if (a == aggregate) {
216                     originalAggregateFound = true;
217                     return;
218-                }        
219+                }
220             });
221             if (originalAggregateFound) {
222                 continue;
223
224
```

**Comment by b-scholz** [comment]
You have spaces after the closing curly brace that need to be removed.

File: src/ast/transform/MaterializeAggregationQueries.cpp (hunk 7 above)
Comment (rdowavic) [reply src/ast/transform/MaterializeAggregationQueries.cpp:109]: what is ownership?

Comment (rdowavic) [reply src/ast/transform/MaterializeAggregationQueries.cpp:109]: I suppose I should be doing it but I really feel like I've run out of time at this point.

Comment (mmcgr) [reply src/ast/transform/MaterializeAggregationQueries.cpp]: Ownership of the TODO - some record of who is responsible or what the issue is.
```
// TODO (rdowavic): Deal with ...
```
or
```
// TODO (#<issuenum>): Deal with ...
```

Comment (mmcgr) [reply src/ast/transform/MaterializeAggregationQueries.cpp]: Weird. The changes weren't showing up, but commenting on it made them show up immediately. All good now.

**Review by mmcgr — APPROVED** [review]
